pybind11/how-to.rst

```diff
@@ -650,105 +771,40 @@ Following :ref:`this rule <style-guide-pybind11-declare-template-wrappers>` we d
-^^^^^^^^^^^^^^^^^^^^
-
-The import statement:
-
-.. code-block:: cpp
-
-    py::module::import("exampleTwo");
-
-in the previous example is present because ``ExampleThree`` depends on ``ExampleBase`` which is defined in a different module (i.e. ``exampleTwo``).
-
-Thus, if you forget to add the import statement, the type ``ExampleBase`` is unknown if ``exampleThree`` happens to be imported before ``exampleTwo``.
-
-Adding these import statements in the C++ wrapper, rather than relying on import order in ``__init__`` in Python, follows :ref:`our rule on import <style-guide-pybind11-import>`.
 
-Finished wrappers (C++ part)
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Cross-module dependencies
+^^^^^^^^^^^^^^^^^^^^^^^^^
 
-The end results for the C++ part of the wrappers (see next for the Python part) are ``exampleTwo.cc``:
+All of the depenxencies in the example classes we've defined here are within the same compiled module, and hence we've been able to rely on ``WrapperCollection`` and its callback system to ensure the wrappers are defined in an order that works.
+The only exception is inheritance: ``ExampleTwo`` and ``ExampleThree`` both inherit from ``ExampleBase``, and that means it's critical that the ``wrapType`` call (or more precisely, the ``py::class_`` instantiation) for ``ExampleBase`` appears before that of either of its derived classes.
 
-.. code-block:: cpp
-
-    #include "pybind11/pybind11.h"
-
-    #include "lsst/TMPL/ExampleTwo.h"
+When dependencies extend beyond module boundaries, we need to import the modules that provide the classes we're using.
+While it's possible to do that directly with ``py::module::import`` calls, it's more readable and less error-prone to use the ``addInheritanceDependency`` and ``addSignatureDependency`` methods of ``WrapperCollection``.
+As its name suggests, the former is needed when you want to inherit from a base class defined in another module; this must be called before the ``py::class_`` instantiation of the derived class.
+``addSignatureDependency`` declares that the external type is used only in function or method signatures.
+Regardless of when it is called, these dependencies will be imported after all local types are declared and before any definition callbacks are run.
 
-    namespace py = pybind11;
-    using namespace pybind11::literals;
-
-    namespace lsst {
-    namespace tmpl {
+Circular *inheritance* dependencies are impossible in both C++ and Python, but circular signatures dependencies are relatively common in C++ (via forward declarations) and quite possible in Python (via function-level imports and duck-typing).
+In pybind11 wrappers, circular signature dependencies are a problem.
+Using ``WrapperCollection`` solves that problem within a single module (in which all wrappers are added to the same ``WrapperCollection`` instance), and it takes steps to solve across multiple modules, by executing signature-dependency imports between type definitions and function definitions.
```

Comment:
"takes steps to solve" -> "partially solves this"?

Reply:
Maybe I should just delete the second half of this sentence entirely, at least for now; I'm sort of speculating about what Python does when it sees circular imports, and hence I'm definitely speculating on whether `WrapperCollection` will actually make that work.

Reply:
I've reworded this to just say that `WrapperCollection` does its best to solve circular dependencies across modules, but that this relies on how Python itself handles circular dependencies and there are no guarantees.